lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -994,38 +1005,71 @@ swift::castValueToABICompatibleType(SILBuilder *builder, SILLocation loc,
 
     // Unwrap the original optional value.
     auto *someDecl = builder->getASTContext().getOptionalSomeDecl();
-    auto *noneBB = builder->getFunction().createBasicBlock();
-    auto *someBB = builder->getFunction().createBasicBlock();
     auto *curBB = builder->getInsertionPoint()->getParent();
-
     auto *contBB = curBB->split(builder->getInsertionPoint());
-    contBB->createPhiArgument(destTy, OwnershipKind::Owned);
+    auto *someBB = builder->getFunction().createBasicBlockAfter(curBB);
+    auto *noneBB = builder->getFunction().createBasicBlockAfter(someBB);
+
+    auto *phi = contBB->createPhiArgument(destTy, value.getOwnershipKind());
+    if (phi->getOwnershipKind() == OwnershipKind::Guaranteed) {
+      auto createEndBorrow = [&](SILBasicBlock::iterator insertPt) {
+        builder->setInsertionPoint(insertPt);
+        builder->createEndBorrow(loc, phi);
+      };
+      for (SILInstruction *user : usePoints) {
+        if (isa<TermInst>(user)) {
+          assert(!isa<BranchInst>(user) && "no branch as guaranteed use point");
+          for (auto *succBB : user->getParent()->getSuccessorBlocks()) {
+            createEndBorrow(succBB->begin());
+          }
+          continue;
+        }
+        createEndBorrow(std::next(user->getIterator()));
+      }
+    }
 
     SmallVector<std::pair<EnumElementDecl *, SILBasicBlock *>, 1> caseBBs;
     caseBBs.push_back(std::make_pair(someDecl, someBB));
     builder->setInsertionPoint(curBB);
     builder->createSwitchEnum(loc, value, noneBB, caseBBs);
-
-    // Handle the Some case.
-    builder->setInsertionPoint(someBB);
-    SILValue unwrappedValue =
-        builder->createUncheckedEnumData(loc, value, someDecl);
+    // In OSSA switch_enum destinations have terminator results.
+    //
+    // TODO: This should be in a switchEnum utility.
+    SILValue unwrappedValue;
+    if (builder->hasOwnership()) {
+      // Create a terminator result, NOT a phi, despite the API name.
+      noneBB->createPhiArgument(value->getType(), OwnershipKind::None);
+      unwrappedValue =
+        someBB->createPhiArgument(optionalSrcTy, value.getOwnershipKind());
+    } else {
+      builder->setInsertionPoint(someBB);
+      unwrappedValue = builder->createUncheckedEnumData(loc, value, someDecl);
+    }
     // Cast the unwrapped value.
+    builder->setInsertionPoint(someBB);
```

Comment:
Isn't this wrong? I think you probably want to hoist this into the hasOwnership section. Since if you add a use with this builder for unwrappedValue, I think you will insert a value before unwrappedValue, breaking dominance.

Reply:
Thanks, I'm sure I meant to do that. I tried to add unit tests for all paths in the OSSA case, but there might still be some holes in the non-OSSA cases.

This whole function is kind of a mess and should be improved with common utilities. These fixes are a stop-gap since the code was already in use outside of simplify-cfg.